File: platform/macos/monarch_macos/Logger.swift
//
//  Logger.swift
//  monarch
//
//  Created by Fernando Trigoso on 5/6/20.
//  Copyright © 2020 Fernando Trigoso. All rights reserved.
//

import Foundation

func printMessage(_ message: String) {
    print("mac_app: \(message)")
}

enum LogLevel : Int {
    case ALL = 0
    case FINEST = 300
    case FINER = 400
    case FINE = 500
    case CONFIG = 700
    case INFO = 800
    case WARNING = 900
    case SEVERE = 1000
    case SHOUT = 1200
    case OFF = 2000
}

func logLevelFromString(levelString: String) -> LogLevel {
    switch levelString {
    case "ALL":
        return LogLevel.ALL
    case "FINEST":
        return LogLevel.FINEST
    case "FINER":
        return LogLevel.FINER
    case "FINE":
        return LogLevel.FINE
    case "CONFIG":
        return LogLevel.CONFIG
    case "INFO":
        return LogLevel.INFO
    case "WARNING":
        return LogLevel.WARNING
    case "SEVERE":
        return LogLevel.SEVERE
    case "SHOUT":
        return LogLevel.SHOUT
    case "OFF":
        return LogLevel.OFF
    default:
        return LogLevel.ALL
    }
}

func logLevelToString(level: LogLevel) -> String {
    switch level {
    case LogLevel.ALL:
        return "ALL";
    case LogLevel.FINEST:
        return "FINEST"
    case LogLevel.FINER:
        return "FINER"
    case LogLevel.FINE:
        return "FINE"
    case LogLevel.CONFIG:
        return "CONFIG"
    case LogLevel.INFO:
        return "INFO"
    case LogLevel.WARNING:
        return "WARNING"
    case LogLevel.SEVERE:
        return "SEVERE"
    case LogLevel.SHOUT:
        return "SHOUT"
    case LogLevel.OFF:
        return "OFF"
    }
}

var defaultLogLevel: LogLevel = LogLevel.ALL


class Logger {
    let name: String
    
    var _level: LogLevel?
    
    var level: LogLevel {
        get {
            return _level ?? defaultLogLevel
        }
        set (newLevel) {
            _level = newLevel
        }
    }
    
    init(_ name: String) {
        self.name = name
    }
    
    func isLoggable(_ value: LogLevel) -> Bool {
        return value.rawValue >= level.rawValue
    }
    
    func log(_ logLevel: LogLevel, _ message: String) {
        if (isLoggable(logLevel)) {
            let levelString = logLevelToString(level: logLevel)
            printMessage("\(levelString) [\(name)] \(message)")
        }
    }
    
    func finest(_ message: String) {
        self.log(LogLevel.FINEST, message);
    }

    func finer(_ message: String) {
        self.log(LogLevel.FINER, message);
    }

    func fine(_ message: String) {
        self.log(LogLevel.FINE, message);
    }

    func config(_ message: String) {
        self.log(LogLevel.CONFIG, message);
    }

    func info(_ message: String) {
        self.log(LogLevel.INFO, message);
    }

    func warning(_ message: String) {
        self.log(LogLevel.WARNING, message);
    }

    func severe(_ message: String) {
        self.log(LogLevel.SEVERE, message);
    }

    func shout(_ message: String) {
        self.log(LogLevel.SHOUT, message);
    }

}

File: platform/windows/src/logger.cpp
#include "logger.h"
#include <iostream>

using namespace std::literals::string_literals;


LogLevel defaultLogLevel = LogLevel::ALL;

LogLevel logLevelFromString(const std::string& levelString) {
    if (levelString == "ALL") {
        return LogLevel::ALL;
    }
    if (levelString == "FINEST") {
        return LogLevel::FINEST;
    }
    if (levelString == "FINER") {
        return LogLevel::FINER;
    }
    if (levelString == "FINE") {
        return LogLevel::FINE;
    }
    if (levelString == "CONFIG") {
        return LogLevel::CONFIG;
    }
    if (levelString == "INFO") {
        return LogLevel::INFO;
    }
    if (levelString == "WARNING") {
        return LogLevel::WARNING;
    }
    if (levelString == "SEVERE") {
        return LogLevel::SEVERE;
    }
    if (levelString == "SHOUT") {
        return LogLevel::SHOUT;
    }
    if (levelString == "OFF") {
        return LogLevel::OFF;
    }
    return LogLevel::ALL;
}

std::wstring logLevelToString(const LogLevel& level) {
    switch (level)
    {
    case LogLevel::ALL: {
        return L"ALL";
    }
    case LogLevel::FINEST: {
        return L"FINEST";
    }
    case LogLevel::FINER: {
        return L"FINER";
    }
    case LogLevel::FINE: {
        return L"FINE";
    }
    case LogLevel::CONFIG: {
        return L"CONFIG";
    }
    case LogLevel::INFO: {
        return L"INFO";
    }
    case LogLevel::WARNING: {
        return L"WARNING";
    }
    case LogLevel::SEVERE: {
        return L"SEVERE";
    }
    case LogLevel::SHOUT: {
        return L"SHOUT";
    }
    case LogLevel::OFF: {
        return L"OFF";
    }
    default:
        return L"UNKNOWN";
    }
}

LogLevel Logger::getLevel() {
    return Logger::_level;
}

void Logger::setLevel(LogLevel newLevel) {
    Logger::_level = newLevel;
}

Logger::Logger(const std::wstring& name) {
    Logger::name = name;
}

bool Logger::isLoggable(LogLevel value) {
    return value >= Logger::_level;
}

void Logger::log(LogLevel logLevel, const std::string& message) {
    Logger::log(logLevel, std::wstring(message.begin(), message.end()));
}

void Logger::log(LogLevel logLevel, const std::wstring& message) {
    if (Logger::isLoggable(logLevel)) {
        std::wstring levelString(logLevelToString(logLevel));
        std::wcout << windows_app << levelString << L" [" << Logger::name << L"] " << message << std::endl;
    }
}


void Logger::finest(const std::string& message) {
    Logger::log(LogLevel::FINEST, message);
}
void Logger::finer(const std::string& message) {
    Logger::log(LogLevel::FINER, message);
}
void Logger::fine(const std::string& message) {
    Logger::log(LogLevel::FINE, message);
}
void Logger::config(const std::string& message) {
    Logger::log(LogLevel::CONFIG, message);
}
void Logger::info(const std::string& message) {
    Logger::log(LogLevel::INFO, message);
}
void Logger::warning(const std::string& message) {
    Logger::log(LogLevel::WARNING, message);
}
void Logger::severe(const std::string& message) {
    Logger::log(LogLevel::SEVERE, message);
}
void Logger::shout(const std::string& message) {
    Logger::log(LogLevel::SHOUT, message);
}


void Logger::finest(const std::wstring& message) {
    Logger::log(LogLevel::FINEST, message);
}
void Logger::finer(const std::wstring& message) {
    Logger::log(LogLevel::FINER, message);
}
void Logger::fine(const std::wstring& message) {
    Logger::log(LogLevel::FINE, message);
}
void Logger::config(const std::wstring& message) {
    Logger::log(LogLevel::CONFIG, message);
}
void Logger::info(const std::wstring& message) {
    Logger::log(LogLevel::INFO, message);
}
void Logger::warning(const std::wstring& message) {
    Logger::log(LogLevel::WARNING, message);
}
void Logger::severe(const std::wstring& message) {
    Logger::log(LogLevel::SEVERE, message);
}
void Logger::shout(const std::wstring& message) {
    Logger::log(LogLevel::SHOUT, message);
}
